h5py/_hl/attrs.py

```diff
@@ -116,10 +116,18 @@ def create(self, name, data, shape=None, dtype=None):
 
         with phil:
 
-            # First, make sure we have a NumPy array.  We leave the data
-            # type conversion for HDF5 to perform.
+            # First, make sure we have a NumPy array.  We leave the data type
+            # conversion for HDF5 to perform (other than the below exception).
             if not isinstance(data, Empty):
+                # Whether we were already given a numpy array or whether we are
+                # converting to a numpy array ourselves:
+                auto_convert_to_array = not isinstance(data, numpy.ndarray)
```

Comment:
Just a brief thought: is this check sufficient? There may be other array-like things (such as h5py datasets, or pandas series) which aren't an ndarray, but have similar constraints. Maybe we should more explicitly check for a list/tuple of strings?

Reply:
I was aiming to support setting as wide a range of things as possible: to auto-convert anything that is not a numpy array, and that, when converted to a numpy array, would have the numpy `U` datatype. I wouldn't want to limit it to just strings and tuples because there are other iterables, including custom ones that are not subclasses of the builtins. I figure passing in anything that is not a numpy array or a subclass of a numpy array is safe to treat as consent to data types being changed. But I might be wrong: I suppose things that it is arguable shouldn't be auto converted are types that are sufficiently numpy-array-like that their datatype should be considered a deliberate choice and respected (where 'respected' just means 'error if it's `U` type instead of silent conversion').

Looks like you can make a pandas Series with vlen str dtype:

```pandas.Series(np.array(['a', 'b'], dtype=h5py.special_dtype(vlen=str)))```

and this can be saved as a dataset or attribute just fine, with or without this patch.

But pandas series don't seem to support the numpy `U` type though ([bug report](https://github.com/pandas-dev/pandas/issues/5261)), so at the moment a user cannot try to use a pandas Series to intentionally write an array with `U` dtype. So there is no risk (yet) that they will get silent conversion instead of an error.

I think there's no need to exclude h5py datasets though because we know for a fact that these, when converted to numpy arrays, will not have `U` datatype (since HDF5 has no such datatype).

If we want to be paranoid about the future of pandas, if they decide to support the `U` type, we could put in something like `if isinstance(getattr(data, 'dtype', None), np.dtype) and ...` instead of the check for `isinstance(data, numpy.ndarray)`. But I suspect this is getting ahead of ourselves, and would basically just be for pandas since other classes out there that have a numpy dtype but not as an attribute called `.dtype`.

What other things are there out there that have numpy datatypes (and therefore might have the `U` datatype), but aren't actual numpy arrays or subclasses? I'm guessing few enough to not be not worth worrying about.

All in all I think it's not too bad for h5py to demand that if you want your datatype respected (again, where that just means error instead of silent conversion), you need to pass in a numpy array. 

I could change the documentation though to make it sound more like it's about anything that isn't a numpy array. It says that in the text, but the title definitely makes it sound like it's about only lists and tuples.

I'm just thinking out loud, but I'll make any change you want really!

Reply:
My rationale for wanting to restrict it more is twofold:

1. Lots of things are iterable, but it doesn't necessarily make sense to store them as attributes, especially if they won't roundtrip. If I try to store some container object with special properties, I'd rather it just refused than quietly converted it to something else. In this particular context, I think duck-typing is more of a hindrance than a help.
2. If we we allow only selected types (lists, tuples) for now, it's easy to add more later when a need becomes apparent. It's much trickier to restrict an initially broad category later if problems come up, because that might break code already in use. So my instinct is to start with a narrow list of types.

It's not my project, though - I've only just joined as a maintainer. I'd like @tacaswell or @aragilar to weigh in on this.

Reply:
I agree we should only do this on tuples/lists only, this is more of a convenience than anything else, as with strings and h5py the best thing to do is to do the conversion beforehand (so that you know what is written and going to be read out).